ipld/hamt/src/node.rs

```diff
@@ -228,48 +241,79 @@ where
         key: K,
         value: V,
         store: &S,
-    ) -> Result<(), Error> {
+        overwrite: bool,
+    ) -> Result<(Option<V>, bool), Error>
+    where
+        V: PartialEq,
+    {
         let idx = hashed_key.next(bit_width)?;
 
         // No existing values at this point.
         if !self.bitfield.test_bit(idx) {
             self.insert_child(idx, key, value);
-            return Ok(());
+            return Ok((None, true));
         }
 
         let cindex = self.index_for_bit_pos(idx);
         let child = self.get_child_mut(cindex);
 
         match child {
             Pointer::Link { cid, cache } => {
-                let cached_node = std::mem::take(cache);
-                let mut child_node = if let Some(sn) = cached_node.into_inner() {
-                    sn
-                } else {
+                cache.get_or_try_init(|| {
                     store
                         .get(cid)?
-                        .ok_or_else(|| Error::CidNotFound(cid.to_string()))?
-                };
-
-                child_node.modify_value(hashed_key, bit_width, depth + 1, key, value, store)?;
-                *child = Pointer::Dirty(child_node);
-                Ok(())
-            }
-            Pointer::Dirty(n) => {
-                Ok(n.modify_value(hashed_key, bit_width, depth + 1, key, value, store)?)
+                        .ok_or_else(|| Error::CidNotFound(cid.to_string()))
+                })?;
+                let child_node = cache.get_mut().expect("filled line above");
+
+                let (old, modified) = child_node.modify_value(
+                    hashed_key,
+                    bit_width,
+                    depth + 1,
+                    key,
+                    value,
+                    store,
+                    overwrite,
+                )?;
+                if modified {
+                    *child = Pointer::Dirty(std::mem::take(child_node));
+                }
+                Ok((old, modified))
             }
+            Pointer::Dirty(n) => Ok(n.modify_value(
+                hashed_key,
+                bit_width,
+                depth + 1,
+                key,
+                value,
+                store,
+                overwrite,
+            )?),
             Pointer::Values(vals) => {
                 // Update, if the key already exists.
                 if let Some(i) = vals.iter().position(|p| p.key() == &key) {
-                    vals[i].1 = value;
-                    return Ok(());
+                    if overwrite && vals[i].value() != &value {
+                        // Can overwrite, replace and return true.
+                        return Ok((Some(std::mem::replace(&mut vals[i].1, value)), true));
+                    } else {
+                        // Can't overwrite, return None and false that the Node was not modified.
+                        return Ok((None, false));
+                    }
```

Comment:
@austinabell Isn't this where you were going to address that additional case you mentioned on our review call? Also, can you describe why that came to mind, so I can understand better?

Reply:
So the point I was thinking through is not relevant to consensus, because the go implementation does not return the previous value. However, thinking through it a third time, I think that in this case a few lines above should be modified a bit. 

The case is on set (overwrite = true) and the values are equal, should the value being passed in be returned back. Although it seems weird internally to do this, it would match a `HashMap` functionality by returning the value that existed. The difference would be that we are not swapping the values, and assuming PartialEq is implemented correctly. I am going to switch it to returning the value that existed, to match a `HashMap` implementation, but open to just returning the value passed in. This operation should be super quick in every case so I think it's the best decision

Reply:
Also, just to be clear, the only case which this would have unintended behaviour is when `PartialEq` doesn't match serialized bytes equality `AND` this was the only value changed, the parent roots won't be updated. Definitely nothing to worry about, but worth keeping in mind that `PartialEq` must match bytes equality to interop in every case with go.